File: packages/expo-rust-jsi/android/src/main/java/expo/modules/rustjsi/ExpoRustJsiModule.kt
package expo.modules.rustjsi

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Expo module that loads and initializes Rust-based JSI modules.
 *
 * The install is exposed as an explicit Function rather than using OnCreate,
 * because OnCreate fires during module instantiation — before the JSI runtime
 * is guaranteed to be available. The JS layer calls install() synchronously
 * on import.
 */
class ExpoRustJsiModule : Module() {
    companion object {
        init {
            System.loadLibrary("expo-rust-jsi")
        }
    }

    override fun definition() = ModuleDefinition {
        Name("ExpoRustJsi")

        Function("install") {
            val runtime = appContext.runtime
            nativeInstall(runtime.rawPointer)
        }
    }

    private external fun nativeInstall(jsiRuntimePtr: Long)
}

File: packages/expo-rust-jsi/ios/ExpoRustJsiModule.swift
import ExpoModulesCore

/// Expo module that loads and initializes Rust-based JSI modules on iOS.
///
/// The install is exposed as an explicit Function rather than using OnCreate,
/// because OnCreate fires during module instantiation — before the JSI runtime
/// is set on AppContext. The JS layer calls install() synchronously on import.
public class ExpoRustJsiModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoRustJsi")

        Function("install") {
            guard let appContext else {
                throw Exceptions.AppContextLost()
            }
            let runtime = try appContext.runtime
            ExpoRustJsiInstall(runtime)
        }
    }
}
